### PR TITLE
no longer turn \n into n when fetching stuff via json_app_get

### PR DESCRIPTION
**Pull request by VVelox**
Came across this while working on Sneck support.

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by murrant** [comment]
why are we running stripslashes() in the first place?

**Comment by VVelox** [comment]
That is actually a very good question, dude. I seem to vaguely remember it may of had some purpose at some point a long time ago, it's been close to half a decade I wrote this original code.

Going to test it on my system and see if it yacks.

**Review by murrant — DISMISSED** [review]

**Comment by VVelox** [comment]
Well I yanked it from it, ran sneck -u after shoving in an something to fail, and it came back as expected.

Now to see if anything breaks when the tests finish. For some reason I vaguely remember netsnmpd a long time ago doing something funny with some characters and breaking JSON, but I may be totally wrong there.

**Comment by VVelox** [comment]
Well it broke linux_chronyd. Will have to look into that and see why.

**Comment by murrant** [comment]
`lnms dev:simulate -o linux_chronyd` (or whatever it is) to simulate as a device added to LibreNMS.

**Comment by Jellyfrog** [comment]
snmpget seems to escape the output to this:
```
"{\"data\": {\"tracking\": {\"reference_name\": \"50505300\", \"reference_type\": \"PPS\", \"stratum\": \"1\", \"reference_time\": \"1613135374.765364947\", \"system_time\": \"-0.000000016\", \"last_offset\": \"0.000000015\", \"rms_offset\": \"0.000000095\", \"frequency\": \"-9.845\", \"residual_frequency\": \"0.000\", \"skew\": \"0.006\", \"root_delay\": \"0.000000001\", \"root_dispersion\": \"0.000008054\", \"update_interval\": \"8.0\", \"leap_status\": \"Normal\"}, \"sources\": [{\"source_mode\": \"#\", \"source_state\": \"*\", \"source_name\": \"PPS\", \"stratum\": \"0\", \"polling_rate\": \"3\", \"reachability\": \"377\", \"last_rx\": \"7\", \"adjusted_offset\": \"0.000000045\", \"measured_offset\": \"0.000000058\", \"estimated_error\": \"0.000000232\", \"number_samplepoints\": \"19\", \"number_runs\": \"10\", \"span\": \"144\", \"frequency\": \"0.000\", \"frequency_skew\": \"0.006\", \"offset\": \"0.000000000\", \"stddev\": \"0.000000271\"}, {\"source_mode\": \"#\", \"source_state\": \"-\", \"source_name\": \"GPS\", \"stratum\": \"0\", \"polling_rate\": \"3\", \"reachability\": \"377\", \"last_rx\": \"7\", \"adjusted_offset\": \"-0.012771881\", \"measured_offset\": \"-0.012771881\", \"estimated_error\": \"0.003427120\", \"number_samplepoints\": \"15\", \"number_runs\": \"9\", \"span\": \"113\", \"frequency\": \"-81.117\", \"frequency_skew\": \"218.921\", \"offset\": \"-0.005838305\", \"stddev\": \"0.006443196\"}]}, \"error\": false, \"errorString\": \"\", \"version\": 1}"
```

So without `stripslashes()` it can't be parsed.
But since stripslashes on removes one slash, and snmpget adds another `\n` should work.

So for example, given the input data: `{"data": {"trac\nking"...`
* snmpget returns: `"{\"data\": {\"trac\\nking\"...`
* stripslashes returns: `"{"data": {"trac\nking"...`
* json_decode returns:
```
array:4 [
  "data" => array:2 [
    "trac\nking" => array:14 [
```

So are you sure this is where the problem comes from?

**Comment by Jellyfrog** [comment]
Even more annoyingly, the data received by net-snmp isnt escaped:

```
Received 1369 byte packet from UDP: [127.1.6.1]:1161->[0.0.0.0]:37117
0000: 30 82 05 55  02 01 01 04  0D 6C 69 6E  75 78 5F 63    0..U.....linux_c
0016: 68 72 6F 6E  79 64 A2 82  05 3F 02 04  2D DB 37 3E    hronyd...?..-.7>
0032: 02 01 00 02  01 00 30 82  05 2F 30 82  05 2B 06 15    ......0../0..+..
0048: 2B 06 01 04  01 BF 08 01  03 02 03 01  02 07 63 68    +.............ch
0064: 72 6F 6E 79  64 04 82 05  10 7B 22 64  61 74 61 22    ronyd....{"data"
0080: 3A 20 7B 22  74 72 61 63  5C 6E 6B 69  6E 67 22 3A    : {"trac\nking":
0096: 20 7B 22 72  65 66 65 72  65 6E 63 65  5F 6E 61 6D     {"reference_nam
0112: 65 22 3A 20  22 35 30 35  30 35 33 30  30 22 2C 20    e": "50505300",
0128: 22 72 65 66  65 72 65 6E  63 65 5F 74  79 70 65 22    "reference_type"
0144: 3A 20 22 50  50 53 22 2C  20 22 73 74  72 61 74 75    : "PPS", "stratu
```
its the client adding it, and can't be turned off it seems.

**Comment by VVelox** [comment]
> snmpget seems to escape the output to this:
> 
> ```
> "{\"data\": {\"tracking\": {\"reference_name\": \"50505300\", \"reference_type\": \"PPS\", \"stratum\": \"1\", \"reference_time\": \"1613135374.765364947\", \"system_time\": \"-0.000000016\", \"last_offset\": \"0.000000015\", \"rms_offset\": \"0.000000095\", \"frequency\": \"-9.845\", \"residual_frequency\": \"0.000\", \"skew\": \"0.006\", \"root_delay\": \"0.000000001\", \"root_dispersion\": \"0.000008054\", \"update_interval\": \"8.0\", \"leap_status\": \"Normal\"}, \"sources\": [{\"source_mode\": \"#\", \"source_state\": \"*\", \"source_name\": \"PPS\", \"stratum\": \"0\", \"polling_rate\": \"3\", \"reachability\": \"377\", \"last_rx\": \"7\", \"adjusted_offset\": \"0.000000045\", \"measured_offset\": \"0.000000058\", \"estimated_error\": \"0.000000232\", \"number_samplepoints\": \"19\", \"number_runs\": \"10\", \"span\": \"144\", \"frequency\": \"0.000\", \"frequency_skew\": \"0.006\", \"offset\": \"0.000000000\", \"stddev\": \"0.000000271\"}, {\"source_mode\": \"#\", \"source_state\": \"-\", \"source_name\": \"GPS\", \"stratum\": \"0\", \"polling_rate\": \"3\", \"reachability\": \"377\", \"last_rx\": \"7\", \"adjusted_offset\": \"-0.012771881\", \"measured_offset\": \"-0.012771881\", \"estimated_error\": \"0.003427120\", \"number_samplepoints\": \"15\", \"number_runs\": \"9\", \"span\": \"113\", \"frequency\": \"-81.117\", \"frequency_skew\": \"218.921\", \"offset\": \"-0.005838305\", \"stddev\": \"0.006443196\"}]}, \"error\": false, \"errorString\": \"\", \"version\": 1}"
> ```
> 
> So without `stripslashes()` it can't be parsed. But since stripslashes on removes one slash, and snmpget adds another; `\n` should work.
> 
> So for example, given the input data: `{"data": {"trac\nking"...`
> 
>     * snmpget returns: `"{\"data\": {\"trac\\nking\"...`
> 
>     * stripslashes returns: `"{"data": {"trac\nking"...`
> 
>     * json_decode returns:
> 
> 
> ```
> array:4 [
>   "data" => array:2 [
>     "trac\nking" => array:14 [
> ```
> 
> So are you sure this is where the problem comes from?

Ahh, yeah! That is along the lines of what I remember it doing.

Not had time to sit down and extract it, but I am guessing it has those slashes in it.

And yeah, this is where the issue comes from. It appears that it is not always getting escaped when returned, depending on the system.

**Comment by VVelox** [comment]
I can say the escaping issue has been fixed on FreeBSD 12-stable when it comes to the stuff in the ports. Likely should not rely on that else where as Jellyfrog has come back across there.

So I've dropped it back into place. Also included a rand value in the replacement key as well now.

**Comment by Jellyfrog** [comment]
This solution is way to hackish and I don't understand what the rand() adds. Either create a string with something like null bytes, or better, replace the stripslashes with a regex that ignore new lines. Best solution would be if net-snmp would be consistent...

**Comment by murrant** [comment]
It is the display hint adding slashes I think.  If you use the MIB, it probably won't have the slashes.

**Comment by VVelox** [comment]
> This solution is way to hackish and I don't understand what the rand() adds. Either create a string with something like null bytes, or better, replace the stripslashes with a regex that ignore new lines. Best solution would be if net-snmp would be consistent...

I would argue that regex is way more hackish as relies on unneeded complexity where there is not a need one one. This is just a very quick and simple replacement, no need for checking backwards.

The point of using rand is to make sure the replacement key is not something predictable. And the use of something long and not likely to show up makes it less likely that something will accidentally get mangled.

No different than using a large number of null bytes.

**Comment by VVelox** [comment]
> It is the snmp oid display hint adding slashes I think. If you use the MIB, it probably won't have the slashes.

Aye. Need to look into this later and see how repeatable getting it not to do that is across systems.

**Comment by VVelox** [comment]
@Jellyfrog Went and replaced the readable strings on the end of the replacement key with three null bytes.

Left the rand() in there just to make sure it is far less likely to be accidentally hit on.

**Comment by VVelox** [comment]
@Jellyfrog Ahh! You meant the entire thing. Yeah, that works just as well.

**Comment by Jellyfrog** [comment]
I dont understand how this works atm.
especially not how it handles escaped quotes.


Maybe something like `str_replace(["\\", "\\"", "\'"], ["\", "\"", "'"], $input);` is enough.

- `\\` > `\`
- `\"` > `"`
- `\'` > `'`

**Comment by Jellyfrog** [comment]
I dont understand how this works atm.
especially not how it handles escaped quotes.


Maybe something like `str_replace(["\\", "\\"", "\'"], ["\", "\"", "'"], $input);` is enough.

- `\\` > `\`
- `\"` > `"`
- `\'` > `'`

**Comment by VVelox** [comment]
> I dont understand how this works atm. especially not how it handles escaped quotes.
> 
> Maybe something like `str_replace(["\\", "\\"", "\'"], ["\", "\"", "'"], $input);` is enough.
> 
>     * `\\` > `\`
> 
>     * `\"` > `"`
> 
>     * `\'` > `'`

It does not need to. :)

So `\"` will become `"` and `\\"` will become `\"`. The reason for this is the regexp looks for `\` and then if it finds it, it checks to make sure it is not followed by a `nbfrt\` (`^` inverts the match group), in which case it leaves it alone, but since we are matching against it, we need to copy it, so thus the `()` and and `$1`.

```
$output = '\\n\\b\\f\\r foo \\t\\"\\\'  \\\\" \\\\\' ';

#$output='\{\"da\\\\"ta\": {\"tracking\": {\"reference_name\": \"50505300\", \"reference_type\": \"PPS\", \"stratum\":' .
' \"1\", \"reference_time\": \"1613135374.765364947\", \"system_time\": \"-0.000000016\", \"last_offset\": \"0.00'.
'0000015\", \"rms_offset\": \"0.000000095\", \"frequency\": \"-9.845\", \"residual_frequency\": \"0.000\", \"skew'.
'\": \"0.006\", \"root_delay\": \"0.000000001\", \"root_dispersion\": \"0.000008054\", \"update_interval\": \"8.0'
.'\", \"leap_status\": \"Normal\"}, \"sources\": [{\"source_mode\": \"#\", \"source_state\": \"*\", \"source_name'.
'\": \"PPS\", \"stratum\": \"0\", \"polling_rate\": \"3\", \"reachability\": \"377\", \"last_rx\": \"7\", \"adjus'.
'ted_offset\": \"0.000000045\", \"measured_offset\": \"0.000000058\", \"estimated_error\": \"0.000000232\", \"num'.
'ber_samplepoints\": \"19\", \"number_runs\": \"10\", \"span\": \"144\", \"frequency\": \"0.000\", \"frequency_sk'.
'ew\": \"0.006\", \"offset\": \"0.000000000\", \"stddev\": \"0.000000271\"}, {\"source_mode\": \"#\", \"source_st'.
'ate\": \"-\", \"source_name\": \"GPS\", \"stratum\": \"0\", \"polling_rate\": \"3\", \"reachability\": \"377\", '.
'\"last_rx\": \"7\", \"adjusted_offset\": \"-0.012771881\", \"measured_offset\": \"-0.012771881\", \"estimated_er'.
'ror\": \"0.003427120\", \"number_samplepoints\": \"15\", \"number_runs\": \"9\", \"span\": \"113\", \"frequency'.
'\": \"-81.117\", \"frequency_skew\": \"218.921\", \"offset\": \"-0.005838305\", \"stddev\": \"0.006443196\"}]}, '.
'\"error\": false, \"errorString\": \"\", \"version\": 1}';

echo $output . "\n\n";
echo preg_replace('/\\\([^nbfrt\\\\])/', '$1', $output) . "\n\n";
echo str_replace(["\\", "\\\"", "\'", '\{'], ["\\", "\"", "'", '{'], $output) . "\n";

```

which will give us...

```
\n\b\f\r foo \t\"\'  \\" \\' 

\n\b\f\r foo \t"'  \" \' 

\n\b\f\r foo \t"'  \" \' 
```

Or for the example JSON

```
\{\"da\\"ta\": {\"tracking\": {\"reference_name\": \"50505300\", \"reference_type\": \"PPS\", \"stratum\": \"1\", \"reference_time\": \"1613135374.765364947\", \"system_time\": \"-0.000000016\", \"last_offset\": \"0.000000015\", \"rms_offset\": \"0.000000095\", \"frequency\": \"-9.845\", \"residual_frequency\": \"0.000\", \"skew\": \"0.006\", \"root_delay\": \"0.000000001\", \"root_dispersion\": \"0.000008054\", \"update_interval\": \"8.0\", \"leap_status\": \"Normal\"}, \"sources\": [{\"source_mode\": \"#\", \"source_state\": \"*\", \"source_name\": \"PPS\", \"stratum\": \"0\", \"polling_rate\": \"3\", \"reachability\": \"377\", \"last_rx\": \"7\", \"adjusted_offset\": \"0.000000045\", \"measured_offset\": \"0.000000058\", \"estimated_error\": \"0.000000232\", \"number_samplepoints\": \"19\", \"number_runs\": \"10\", \"span\": \"144\", \"frequency\": \"0.000\", \"frequency_skew\": \"0.006\", \"offset\": \"0.000000000\", \"stddev\": \"0.000000271\"}, {\"source_mode\": \"#\", \"source_state\": \"-\", \"source_name\": \"GPS\", \"stratum\": \"0\", \"polling_rate\": \"3\", \"reachability\": \"377\", \"last_rx\": \"7\", \"adjus
ted_offset\": \"-0.012771881\", \"measured_offset\": \"-0.012771881\", \"estimated_error\": \"0.003427120\", \"number_samplepoints\": \"15\", \"number_runs\": \"9\", \"span\": \"113\", \"fre
quency\": \"-81.117\", \"frequency_skew\": \"218.921\", \"offset\": \"-0.005838305\", \"stddev\": \"0.006443196\"}]}, \"error\": false, \"errorString\": \"\", \"version\": 1}

{"da\"ta": {"tracking": {"reference_name": "50505300", "reference_type": "PPS", "stratum": "1", "reference_time": "1613135374.765364947", "system_time": "-0.000000016", "last_offset": "0.000
000015", "rms_offset": "0.000000095", "frequency": "-9.845", "residual_frequency": "0.000", "skew": "0.006", "root_delay": "0.000000001", "root_dispersion": "0.000008054", "update_interval":
 "8.0", "leap_status": "Normal"}, "sources": [{"source_mode": "#", "source_state": "*", "source_name": "PPS", "stratum": "0", "polling_rate": "3", "reachability": "377", "last_rx": "7", "adj
usted_offset": "0.000000045", "measured_offset": "0.000000058", "estimated_error": "0.000000232", "number_samplepoints": "19", "number_runs": "10", "span": "144", "frequency": "0.000", "freq
uency_skew": "0.006", "offset": "0.000000000", "stddev": "0.000000271"}, {"source_mode": "#", "source_state": "-", "source_name": "GPS", "stratum": "0", "polling_rate": "3", "reachability": 
"377", "last_rx": "7", "adjusted_offset": "-0.012771881", "measured_offset": "-0.012771881", "estimated_error": "0.003427120", "number_samplepoints": "15", "number_runs": "9", "span": "113",
 "frequency": "-81.117", "frequency_skew": "218.921", "offset": "-0.005838305", "stddev": "0.006443196"}]}, "error": false, "errorString": "", "version": 1}

{"da\"ta": {"tracking": {"reference_name": "50505300", "reference_type": "PPS", "stratum": "1", "reference_time": "1613135374.765364947", "system_time": "-0.000000016", "last_offset": "0.00
0000015", "rms_offset": "0.000000095", "frequency": "-9.845", "residual_frequency": "0.000", "skew": "0.006", "root_delay": "0.000000001", "root_dispersion": "0.000008054", "update_interval"
: "8.0", "leap_status": "Normal"}, "sources": [{"source_mode": "#", "source_state": "*", "source_name": "PPS", "stratum": "0", "polling_rate": "3", "reachability": "377", "last_rx": "7", "ad
justed_offset": "0.000000045", "measured_offset": "0.000000058", "estimated_error": "0.000000232", "number_samplepoints": "19", "number_runs": "10", "span": "144", "frequency": "0.000", "fre
quency_skew": "0.006", "offset": "0.000000000", "stddev": "0.000000271"}, {"source_mode": "#", "source_state": "-", "source_name": "GPS", "stratum": "0", "polling_rate": "3", "reachability":
 "377", "last_rx": "7", "adjusted_offset": "-0.012771881", "measured_offset": "-0.012771881", "estimated_error": "0.003427120", "number_samplepoints": "15", "number_runs": "9", "span": "113"
, "frequency": "-81.117", "frequency_skew": "218.921", "offset": "-0.005838305", "stddev": "0.006443196"}]}, "error": false, "errorString": "", "version": 1}
```

The additional replacement is needed there or it will miss the `\{` at the start.

So either works. I just implemented it as a regexp as that is what you suggested.

File: includes/polling/functions.inc.php [hunk 1]
```diff
@@ -634,7 +634,8 @@ function json_app_get($device, $extend, $min_version = 1)
     }
 
     //  turn the JSON into a array
-    $parsed_json = json_decode(stripslashes($output), true);
+    $output = str_replace(['\\', '\\"', "\'", '\{'], ['\\', '"', "'", '{'], $output);
```
Comment (murrant) [inline includes/polling/functions.inc.php:637]: Shouldn't the first one be `'\\\\'` instead of `'\\'`

Comment (VVelox) [reply includes/polling/functions.inc.php]: Yup! Looks like that is correct. Fixed and we will see how the run goes.

```
[kitsune@vixen42]0|/home/kitsune>cat /tmp/test.php
<?php
$a = '\\\\\\"\{'."\'";
echo $a."\n";
$a = str_replace(['\\\\', '\\"', "\'", '\{'], ['\\', '"', "'", '{'], $a);
echo $a."\n";
[kitsune@vixen42]0|/home/kitsune>php /tmp/test.php
\\\"\{\'
\"{'
[kitsune@vixen42]0|/home/kitsune>
```

Comment (VVelox) [reply includes/polling/functions.inc.php]: Woot! All tests passed.

**Review by murrant — APPROVED** [review]

**Review by murrant — CHANGES_REQUESTED** [review]
Hmm, how come this doesn't affect any test data?

**Comment by VVelox** [comment]
> Hmm, how come this doesn't affect any test data?

While a few of items do have SNMP data with escaped characters(chronyd for example), none of them in such a manner that has any data changed by it.

Only one I can think of that would have data affected by this is Sneck, but it's not been committed yet.

**Comment by VVelox** [comment]
hmm.... so found another edge case that does not work nicely with this...

That said #14169 fixes this as well, so going to close this.

**Pull request closed without merging**